File: apps/desktop/src/services/NativeService/browser.ts
import { invoke } from '@tauri-apps/api/core';

import type {
    BrowserActRequest,
    BrowserActResponse,
    BrowserConnectExistingRequest,
    BrowserConnectExistingResponse,
    BrowserExistingSession,
    BrowserInstalledBrowser,
    BrowserNavigationRequest,
    BrowserObserveRequest,
    BrowserObserveResponse,
    BrowserSessionResponse,
    BrowserStartRequest,
    BrowserStatusResponse,
    BrowserTabRequest,
} from './types';

/**
 * Native browser automation bridge.
 */
export const browser = {
    status(): Promise<BrowserStatusResponse> {
        return invoke('browser_status');
    },
    start(request: BrowserStartRequest): Promise<BrowserSessionResponse> {
        return invoke('browser_start', { request });
    },
    discoverInstalled(): Promise<BrowserInstalledBrowser[]> {
        return invoke('browser_discover_installed');
    },
    defaultDataPath(): Promise<string> {
        return invoke('browser_default_data_path');
    },
    discoverExisting(): Promise<BrowserExistingSession[]> {
        return invoke('browser_discover_existing');
    },
    connectExisting(
        request: BrowserConnectExistingRequest
    ): Promise<BrowserConnectExistingResponse> {
        return invoke('browser_connect_existing', { request });
    },
    stop(): Promise<BrowserSessionResponse> {
        return invoke('browser_stop');
    },
    navigate(request: BrowserNavigationRequest): Promise<BrowserSessionResponse> {
        return invoke('browser_navigate', { request });
    },
    back(request: BrowserTabRequest = {}): Promise<BrowserSessionResponse> {
        return invoke('browser_back', { request });
    },
    forward(request: BrowserTabRequest = {}): Promise<BrowserSessionResponse> {
        return invoke('browser_forward', { request });
    },
    reload(request: BrowserTabRequest = {}): Promise<BrowserSessionResponse> {
        return invoke('browser_reload', { request });
    },
    observe(request: BrowserObserveRequest): Promise<BrowserObserveResponse> {
        return invoke('browser_observe', { request });
    },
    act(request: BrowserActRequest): Promise<BrowserActResponse> {
        return invoke('browser_act', { request });
    },
} as const;

File: apps/desktop/src-tauri/src/commands/browser.rs
use tauri::State;

use crate::core::browser::{
    process::{default_managed_browser_data_path, discover_installed_browsers},
    types::{
        BrowserActRequest, BrowserActResult, BrowserConnectExistingRequest,
        BrowserConnectExistingResult, BrowserDescriptor, BrowserExistingSession,
        BrowserNavigateRequest, BrowserObservation, BrowserObserveRequest, BrowserStartRequest,
        BrowserStatus, BrowserTabRequest,
    },
    BrowserRuntime,
};

#[tauri::command]
pub fn browser_status(runtime: State<'_, BrowserRuntime>) -> BrowserStatus {
    runtime.status()
}

#[tauri::command]
pub async fn browser_start(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserStartRequest,
) -> Result<BrowserStatus, String> {
    runtime.start(request).await
}

#[tauri::command]
pub fn browser_discover_installed() -> Vec<BrowserDescriptor> {
    discover_installed_browsers()
}

#[tauri::command]
pub fn browser_default_data_path() -> Result<String, String> {
    default_managed_browser_data_path().map(|path| path.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn browser_discover_existing(
    runtime: State<'_, BrowserRuntime>,
) -> Result<Vec<BrowserExistingSession>, String> {
    runtime.discover_existing_sessions().await
}

#[tauri::command]
pub async fn browser_connect_existing(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserConnectExistingRequest,
) -> Result<BrowserConnectExistingResult, String> {
    runtime.connect_existing(request).await
}

#[tauri::command]
pub fn browser_stop(runtime: State<'_, BrowserRuntime>) -> BrowserStatus {
    runtime.stop()
}

#[tauri::command]
pub async fn browser_navigate(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserNavigateRequest,
) -> Result<BrowserStatus, String> {
    runtime.navigate(request).await
}

#[tauri::command]
pub async fn browser_back(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserTabRequest,
) -> Result<BrowserStatus, String> {
    runtime.history_action(request, "back").await
}

#[tauri::command]
pub async fn browser_forward(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserTabRequest,
) -> Result<BrowserStatus, String> {
    runtime.history_action(request, "forward").await
}

#[tauri::command]
pub async fn browser_reload(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserTabRequest,
) -> Result<BrowserStatus, String> {
    runtime.history_action(request, "reload").await
}

#[tauri::command]
pub async fn browser_observe(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserObserveRequest,
) -> Result<BrowserObservation, String> {
    runtime.observe(request).await
}

#[tauri::command]
pub async fn browser_act(
    runtime: State<'_, BrowserRuntime>,
    request: BrowserActRequest,
) -> Result<BrowserActResult, String> {
    runtime.act(request).await
}
